### PR TITLE
feat: add json_output

### PR DESCRIPTION
**Pull request by Janik-Haag**
This is useful if you have some other tool which makes use of the documentation data, e.g. a function search.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -56,6 +56,10 @@ struct Options {
     #[arg(short, long, default_value_t = String::from("lib"))]
     prefix: String,
 
+    /// Prefix for the category (e.g. 'lib' or 'utils').
+    #[arg(short, long, default_value_t = false)]
+    json_output: bool,
```
Comment (infinisil) [inline src/main.rs:61]:
```suggestion
    /// Whether to output JSON
    #[arg(short, long, default_value_t = false)]
    json_output: bool,
```

File: src/main.rs [hunk 2]
```diff
@@ -316,12 +326,23 @@ fn main() {
...
-            .write_section(&locs, &mut output)
-            .expect("Failed to write section")
+    let entries = collect_entries(nix, &opts.prefix, &opts.category, &locs);
+
+    if opts.json_output {
+        let json_string = match serde_json::to_string(&entries) {
+            Ok(json) => json,
+            Err(error) => panic!("Problem converting entries to json: {error:?}"),
+        };
+        println!("{}", json_string);
```
Comment (infinisil) [inline src/main.rs:336]: This makes it return a JSON list, which might be hard to extend in the future. I'd rather we adhere to [these JSON guidelines](https://github.com/NixOS/nix/blob/master/doc/manual/src/development/json-guideline.md) by returning an object like `{ "functions": [ ... ] }`.

**Review by infinisil — CHANGES_REQUESTED** [review]
Needs at least a test, but other than that and my comments, this looks good to me :)

**Comment by hsjobeki** [comment]
Regarding the whole Format, docbook, etc discussion. I think nixdoc should have a json Output as Default. 
To loosen the coupling between nixdoc and nixpkgs manaul

File: src/main.rs (hunk 2 above)
Comment (Janik-Haag) [reply src/main.rs]: I implemented this as I struct:
```rust
struct JsonFormat {
    version: u32,
    entries: Vec<ManualEntry>,
}
```

**Comment by Janik-Haag** [comment]
> Needs at least a test,

I added a test, and took the opertunity to refactor a bunch of tests, e.g. making the test_main function actually test the main function instead of just some copied code.

> Regarding the whole Format, docbook, etc discussion. I think nixdoc should have a json Output as Default.

I think that this should not happen here, but (if at all) in a new major relase since it breaks current behavior.
Do you see this as a blocker for this pr?

**Review by infinisil — APPROVED** [review]
I don't think the default matters at all really, because I don't expect anybody to use this tool directly, it's all through automation. So definitely not blocking imo, this PR looks good as-is to me, thanks!

**Pull request merged**